src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_readvalue_inputs_to_subgraph.cpp

```diff
@@ -0,0 +1,157 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/cpu_opset/common/op/read_value_with_subgraph.hpp"
+#include "move_readvalue_inputs_to_subgraph.hpp"
+
+#include <unordered_set>
+
+#include "itt.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/pass/constant_folding.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "ov_ops/rotary_positional_embeddings.hpp"
+#include "transformations/rt_info/disable_fp16_compression.hpp"
+#include "transformations/utils/gen_pattern.hpp"
+#include "transformations/utils/utils.hpp"
+#include "transformations/cpu_opset/common/op/submodel.hpp"
+
+ov::intel_cpu::MoveReadValueInputsToSubgraph::MoveReadValueInputsToSubgraph() {
+    MATCHER_SCOPE(MoveReadValueInputsToSubgraph);
+    using namespace ov::pass::pattern;
+
+    auto readvalue_pattern = pass::pattern::wrap_type<ov::op::v6::ReadValue>();
+
+    ov::matcher_pass_callback callback = [=](Matcher& m) {
+        const auto& pattern_map = m.get_pattern_value_map();
+        auto readvalue = as_type_ptr<ov::opset6::ReadValue>(pattern_map.at(readvalue_pattern).get_node_shared_ptr());
+        if (!readvalue || readvalue->get_input_size() != 1u) {
+            return false;
+        }
+
+        int recursive_deep_check_node = 0;
+        constexpr int MAX_RECURSIVE_DEEP_CHECK_NODE = 10;
+        int recursive_deep_check_successor = 0;
+        constexpr int MAX_RECURSIVE_DEEP_CHECK_SUCCESSOR = 10;
+        bool final_successor_is_only_root = true;
+        std::string root_name = readvalue->get_friendly_name();
+
+        NodeVector subgraph = {};
+        auto add_node_to_subgraph = [&](std::shared_ptr<ov::Node> node) {
+            subgraph.emplace_back(node);
+        };
+        NodeVector inputs = {};
+        OutputVector outputs = {};
+
+        std::function<void(std::shared_ptr<ov::Node>)> check_node_successor = [&](std::shared_ptr<ov::Node> node) {
+            recursive_deep_check_successor++;
+            if (recursive_deep_check_successor > MAX_RECURSIVE_DEEP_CHECK_SUCCESSOR) {
+                final_successor_is_only_root = false;
+                return;
+            }
```

Comment:
As far as I understand it, these limitations come from the fact that your algorithm has O(v^2) complexity, where v - is the number of vertices. But, it could be seen as the LCA problem but in the reverse direction (where the outputs of the graph are roots). In that case all we need is to test each node from the preceding subgraph (all the parents of the ReadValue) has a specific LCA with the specific ReadValue node, but in the reverse direction (from parent to child of the original graph). This specific LCA should be the ReadValue itself, otherwise it means that the node from the preceding subgraph is connected to some other node down the graph and may not be included into the init subgraph. The LCA problem may be solved with O(log(v)) complexity, which gives O(v*log(v)) complexity of the whole algo, but at cost of O(v) memory. What do you think?

Reply:
Actually, I think this is different from the LCA problem, LCA is find 2 nodes' lowest common ancestor, my problem is that checking `ReadValue`'s all inputs' final successor still is ReadValue.
And my implementation is very complex, I think I can use a hashmap to reduce some duplicate checking.

Reply:
Added `std::unordered_set<std::shared_ptr<ov::Node>> visited` to flag visited node and avoid repeat visit it.